File: server/observability/tracing.js
/**
 * OpenTelemetry tracing initialization and helpers.
 */

import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { Resource } from '@opentelemetry/resources';
import { SEMRESATTRS_SERVICE_NAME } from '@opentelemetry/semantic-conventions';
import { trace, context, propagation } from '@opentelemetry/api';

const SERVICE_NAME = process.env.OTEL_SERVICE_NAME || 'nexasphere-api';
const OTLP_ENDPOINT = process.env.OTEL_EXPORTER_OTLP_ENDPOINT || 'http://localhost:4318/v1/traces';

let sdk = null;

export function initTracing() {
  if (process.env.OTEL_ENABLED === 'false') {
    return null;
  }

  const exporter = new OTLPTraceExporter({ url: OTLP_ENDPOINT });

  sdk = new NodeSDK({
    resource: new Resource({
      [SEMRESATTRS_SERVICE_NAME]: SERVICE_NAME,
    }),
    traceExporter: exporter,
    instrumentations: [
      getNodeAutoInstrumentations({
        '@opentelemetry/instrumentation-fs': { enabled: false },
      }),
    ],
  });

  sdk.start();

  process.on('SIGTERM', () => {
    sdk?.shutdown().catch(() => {});
  });

  return sdk;
}

export function getActiveTraceId() {
  const span = trace.getSpan(context.active());
  if (!span) return null;
  const ctx = span.spanContext();
  return ctx.traceId && ctx.traceId !== '00000000000000000000000000000000' ? ctx.traceId : null;
}

export function injectTraceHeaders(headers = {}) {
  const carrier = { ...headers };
  propagation.inject(context.active(), carrier);
  return carrier;
}

export { trace, context, propagation };

File: server-python/observability/tracing.py
import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

SERVICE_NAME = os.getenv("OTEL_SERVICE_NAME", "nexasphere-python")
OTLP_ENDPOINT = os.getenv(
    "OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318/v1/traces"
)


def init_tracing(app):
    if os.getenv("OTEL_ENABLED", "true").lower() == "false":
        return

    resource = Resource.create({"service.name": SERVICE_NAME})
    provider = TracerProvider(resource=resource)
    exporter = OTLPSpanExporter(endpoint=OTLP_ENDPOINT)
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)
    FastAPIInstrumentor.instrument_app(app)


def get_active_trace_id() -> str | None:
    span = trace.get_current_span()
    ctx = span.get_span_context()
    if ctx and ctx.trace_id:
        trace_id = format(ctx.trace_id, "032x")
        if trace_id != "0" * 32:
            return trace_id
    return None
